README.md
# Django Init Project

Project template forked on [django-init](https://github.com/Fueled/django-init) for create django projects, optimized for making REST API with deployment on Heroku and EC2 instances via Ansible.

## Requirements

* Python 3.6.x
* Docker
* Docker Compose

## Features

* Django 2.0.x
* Python 3.6.x
* [12-Factor][12factor] based settings management via [django-environ], reads settings from `.env` if present.
* PostresSQL everywhere
* PostGIS support (optional)
* Docker Compose for development (Database, Redis, etc)
* Ready to deploy on Heroku (optional) and Ubuntu 16 LTS via [Ansible](Optional)
* [Django Rest Framework][drf] 3.7.x. ready.
* Uses `django_sites` instead of `django.contrib.sites`
* Use [mkdocs] for project documentation.
* Uses [pytest] as test runner.
* Config files for running isolated tests and deployments to dev/qa/prod environment on Heroku from git branches.
  * Jenkins: `Jenkinsfile`
  * Circle CI: `circle.yml`
  * Travis: `travis.yml`
* Custom `User` app, for easier extensibility.
* Media storage using Amazon S3 (optional)
* [Letsencrypt](https://letsencrypt.org/) Support via [certbot](https://certbot.eff.org)
* SASS with autoprefixing support using django-compressor (optional)
* Livereloading of browser in development via [devrecargar]
* robots.txt and humans.txt configured

## Getting Started

Install cookiecutter with `brew install cookiecutter` or `pip install cookiecutter`.

```
cookiecutter gh:condadolabs/init-project
```

It will ask you couple of questions required to generate the project. It will generate a folder containing all the files in your current working directory.

If you opt to setup the project automatically, it will also:

* initialize a git repo and bump initial tag and version.
* create a virtualenv in the folder `venv` inside the project.
* install all the python dependencies inside it.
* try to create a postgres database and run the initial migration against it, using your container via `docker-compose`.

then only thing you'll need to do is:

1.  `cd` into the new `github_repository` folder just created.
2.  Activate virtualenv `source venv/bin/activate`.
3.  Run `fab serve` or `./manage.py runserver`

Don't forget to carefully look at the generated README. Awesome, right?

You can also explore the [wiki] section for details on advance setup and usages. (SOON)

## Release Policy

`init-project` is a rolling release project. Commit and fixes are added to `master` branch on regular basis and always have latest stable django and associated libraries. You are advised to follow-up with changelogs.

## Changelogs

Refer to [CHANGELOG.md](CHANGELOG.md).

## Code of Conduct

Everyone interacting in the django-init project's codebases, issue trackers, chat rooms, and mailing lists is expected to follow the [PyPA Code of Conduct](https://www.pypa.io/en/latest/code-of-conduct/).

## Related Projects

* https://github.com/Fueled/django-init
* https://github.com/pydanny/cookiecutter-django
* https://github.com/wemake-services/wemake-django-template
* https://github.com/lionheart/django-template

---

Thinks for the [Fueled](https://fueled.com) for creating this template.

[wiki]: https://github.com/condadolabs/init-project/wiki
[mkdocs]: http://www.mkdocs.org/
[12factor]: http://12factor.net
[pytest]: http://pytest.org/
[django-environ]: https://github.com/joke2k/django-environ
[ansible]: http://docs.ansible.com/index.html
[devrecargar]: https://github.com/scottwoodall/django-devrecargar
[drf]: http://www.django-rest-framework.org/


CHANGELOG.md
# ChangeLog
All notable changes to this project will be documented in this file.

## [1.x]

- Project forked on the [django-init](https://github.com/Fueled/django-init)
- Add docker-compose support o create postgresql and redis containers


hooks/post_gen_project.sh
#!/bin/bash
red=`tput setaf 1`
green=`tput setaf 2`
reset=`tput sgr0`

# Ensure newline at EOF
find . ! -path "*/venv/*" -type f -name "*.py" -exec bash -c "tail -n1 {} | read -r _ || echo >> {}" \;

echo "${green}[Finished]${reset}"

echo "==> Setup project dependencies? It will:"
echo ""
echo ""
echo "  - Create virtualenv at './{{ cookiecutter.github_repository }}/venv/'."
echo ""
echo "  - Install development requirements inside virtualenv."
echo ""
echo "  - Run docker-compose to create the containers."
echo ""
echo "  - Run './manage.py migrate'."
echo ""
echo "  - Initialize git."
echo ""
echo "  - Create git tag {{ cookiecutter.version }}."
echo ""
echo ""
echo -n "Would you like to perform these steps? (y/[n]) "
echo ""

# Inside CI, always assume the answer is yes! :)
if [ $CI ]; then
    yn="yes"
else
    read  yn
fi

if echo "{{ cookiecutter.enable_heroku_deployment }}" | grep -iq "^n"; then
    rm -rf uwsgi.ini Procfile runtime.txt bin
fi

if echo "{{ cookiecutter.add_ansible }}" | grep -iq "^n"; then
    rm -rf provisioner Vagrantfile
fi

if echo "{{ cookiecutter.add_sass_with_django_compressor }}" | grep -iq "^n"; then
    rm -rf package.json
else
    mv {{ cookiecutter.main_module }}/static/css/main.css {{ cookiecutter.main_module }}/static/css/main.scss
fi

if echo "{{ cookiecutter.add_celery }}" | grep -iq "^n"; then
    rm -rf {{ cookiecutter.main_module }}/celery.py
fi

if echo "{{ cookiecutter.integration_continuous.lower() }}" | grep -iq "travis"; then
    rm -rf {{ cookiecutter.main_module }}/circle.yml
    rm -rf {{ cookiecutter.main_module }}/Jenkinsfile
fi

if echo "{{ cookiecutter.integration_continuous.lower() }}" | grep -iq "jenkins"; then
    rm -rf {{ cookiecutter.main_module }}/circle.yml
    rm -rf {{ cookiecutter.main_module }}/.travis.yml
fi

if echo "{{ cookiecutter.integration_continuous.lower() }}" | grep -iq "circle"; then
    rm -rf {{ cookiecutter.main_module }}/.travis.yml
    rm -rf {{ cookiecutter.main_module }}/Jenkinsfile
fi

if echo "$yn" | grep -iq "^y"; then
    echo "==> Checking system dependencies. You may need to enter your sudo password."

    echo "==> Install pip, if not present."
    if ! hash pip 2>/dev/null; then
        echo "pip not found.... installing it..."
        sudo easy_install pip
    fi

    echo "==> Install fabric, if not present."
    if ! hash fab 2>/dev/null; then
        echo "fab command not found... installing it..."
        sudo pip install fabric
    fi

    echo "==> Initialize git repo and create first commit and tag it with v{{ cookiecutter.version }}"
    git init
    git add .
    git commit -am "chore(setup): create base django project."
    git tag v{{ cookiecutter.version }}

    echo "${green}==> Setup the project dependencies and database for local development${reset}"
    fab init

    OUT=$?
    if [ $OUT -eq 0 ]; then
        echo "${green}============================================"
        echo "All set! Run these commands to get started:"
        echo ""
        echo "cd {{ cookiecutter.github_repository }}"
        echo "source venv/bin/activate"
        echo "pytest"
        echo "python manage.py runserver"
        echo ""
        echo "============================================"
        echo "${green} ============> HAPPY CODING <============ ${reset}"
    else
        echo "${red}============================================"
        echo "          Oops! Something went wrong!!      "
        echo "============================================${reset}"
        echo ""
        echo -n "HINT: Make sure you have installed all OS dependencies. "
        echo "Check the logs above, they might give you some clues."
    fi
else
    echo "==> Skipping project setup..."
    echo "==> You can now 'cd {{ cookiecutter.github_repository }}/' and explore the project. "
    echo "    Read 'README.md' inside it for further setup instructions!"
    echo ""
    echo "${green} ============> HAPPY CODING <============ ${reset}"
fi


{{cookiecutter.github_repository}}/fabfile.py
# -*- coding: utf-8 -*-
"""Fabric file for managing this project.

See: http://www.fabfile.org/
"""

from __future__ import absolute_import, unicode_literals, with_statement

# Standard Library
import time
from contextlib import contextmanager as _contextmanager
from functools import partial
from os.path import dirname, isdir, join

# Third Party Stuff
from fabric.api import local as fabric_local, env
from fabric import api as fab

local = partial(fabric_local, shell='/bin/bash')


HERE = dirname(__file__)

# ==========================================================================
#  Settings
# ==========================================================================
env.project_name = '{{ cookiecutter.main_module }}'
env.apps_dir = join(HERE, env.project_name)
env.docs_dir = join(HERE, 'docs')
env.virtualenv_dir = join(HERE, 'venv')
env.requirements_file = join(HERE, 'requirements/development.txt')
env.shell = '/bin/bash -l -i -c'

{% if cookiecutter.add_ansible.lower() == 'y' -%}env.use_ssh_config = True
env.dotenv_path = join(HERE, '.env')
env.config_setter = local{% endif %}


def init(vagrant=False):
    """Prepare a local machine for development."""

    install_requirements()
    local('docker-compose up -d')  # creating containers
    time.sleep(5)
    manage('migrate')


def install_requirements(file=env.requirements_file):
    """Install project dependencies."""
    verify_virtualenv()
    # activate virtualenv and install
    with virtualenv():
        local('pip install -r %s' % file)


def serve_docs(options=''):
    """Start a local server to view documentation changes."""
    with fab.lcd(HERE) and virtualenv():
        local('mkdocs serve {}'.format(options))


def shell():
    manage('shell_plus')


def test(options='--pdb --cov'):
    """Run tests locally. By Default, it runs the test using --ipdb.
    You can skip running it using --ipdb by running - `fab test:""`
    """
    with virtualenv():
        local('flake8 .')
        local('pytest %s' % options)


def serve(host='127.0.0.1:8000'):
    """Run local developerment server, making sure that dependencies and
    database migrations are upto date.
    """
    install_requirements()
    migrate()
    manage('runserver %s' % host)
{%- if cookiecutter.add_celery.lower() == 'y' %}


def celery():
    """Run local celery, making sure that dependencies and
    database migrations are upto date.
    """
    install_requirements()
    migrate()
    local('celery worker -A {{ cookiecutter.main_module }} -B -l INFO --concurrency=2')
{%- endif %}


def makemigrations(app=''):
    """Create new database migration for an app."""
    manage('makemigrations %s' % app)


def migrate():
    """Apply database migrations."""
    manage('migrate')


def createapp(appname):
    """fab createapp <appname>
    """
    path = join(env.apps_dir, appname)
    local('mkdir %s' % path)
    manage('startapp %s %s' % (appname, path))
{%- if cookiecutter.add_ansible.lower() == 'y' %}


#  Enviroments & Deployments
# ---------------------------------------------------------
def dev():
    env.host_group = 'dev'
    env.remote = 'origin'
    env.branch = 'master'
    env.hosts = ['dev.{{ cookiecutter.main_module }}.com']
    env.dotenv_path = '/home/ubuntu/dev/{{ cookiecutter.main_module }}/.env'
    env.config_setter = fab.run


def qa():
    env.host_group = 'qa'
    env.remote = 'origin'
    env.branch = 'qa'
    env.hosts = ['qa.{{ cookiecutter.main_module }}.com']
    env.dotenv_path = '/home/ubuntu/qa/{{ cookiecutter.main_module }}/.env'
    env.config_setter = fab.run


def prod():
    env.host_group = 'production'
    env.remote = 'origin'
    env.branch = 'prod'
    env.hosts = ['prod.{{ cookiecutter.main_module }}.com']
    env.dotenv_path = '/home/ubuntu/prod/{{ cookiecutter.main_module }}/.env'
    env.config_setter = fab.run


def config(action=None, key=None, value=None):
    """Read/write to .env file on local and remote machines.

    Usages: fab [prod] config:set,<key>,<value>
            fab [prod] config:get,<key>
            fab [prod] config:unset,<key>
            fab [prod] config:list
    """
    import dotenv
    command = dotenv.get_cli_string(env.dotenv_path, action, key, value)
    env.config_setter('touch %(dotenv_path)s' % env)

    if env.config_setter == local:
        with virtualenv():
            env.config_setter(command)
    else:
        env.config_setter(command)
        restart_servers()


def restart_servers():
    services = ['uwsgi-emperor.service', ]
    for service in services:
        fab.sudo('systemctl restart {0}'.format(service))


def configure(tags='', skip_tags='deploy'):
    """Setup a host using ansible scripts

    Usages: fab [prod|qa|dev] configure
    """
    fab.require('host_group')
    cmd = 'ansible-playbook -i hosts site.yml --limit=%(host_group)s' % env
    with fab.lcd('provisioner'):
        if tags:
            cmd += " --tags '%s'" % tags
        if skip_tags:
            cmd += " --skip-tags '%s'" % skip_tags
        local(cmd)


def deploy():
    configure(tags='deploy', skip_tags='')


def deploy_docs():
    """Deploy documentation to server via ansible.
    """
    configure(tags='documentation', skip_tags=''){% else %}


def deploy_docs():
    """Deploy documentation to github pages.
    """
    with fab.lcd(HERE) and virtualenv():
        local('mkdocs gh-deploy')
        local('rm -rf _docs_html'){% endif %}


# Helpers
# ---------------------------------------------------------
def manage(cmd, venv=True):
    with virtualenv():
        local('python manage.py %s' % cmd)


@_contextmanager
def virtualenv():
    """Activates virtualenv context for other commands to run inside it.
    """
    with fab.cd(HERE):
        with fab.prefix('source %(virtualenv_dir)s/bin/activate' % env):
            yield


def verify_virtualenv():
    """This modules check and install virtualenv if it not present.
    It also creates local virtualenv directory if it's not present
    """
    from distutils import spawn
    if not spawn.find_executable('virtualenv'):
        local('sudo pip install virtualenv')

    if not isdir(env.virtualenv_dir):
        local('virtualenv %(virtualenv_dir)s -p $(which python3)' % env)


{{cookiecutter.github_repository}}/{{cookiecutter.main_module}}/urls.py
"""Root url routering file.

You should put the url config in their respective app putting only a
refernce to them here.
"""

# Third Party Stuff
from django.conf import settings
from django.conf.urls import include, url
from django.conf.urls.static import static
from django.contrib import admin
from django.views.generic import TemplateView

from . import routers
from .base import views as base_views
from .base.api import schemas as api_schemas

admin.site.site_title = admin.site.site_header = '{{ cookiecutter.project_name }} Administration'
handler500 = base_views.server_error

# Top Level Pages
# ==============================================================================
urlpatterns = [
    url(r'^$', TemplateView.as_view(template_name='pages/home.html'), name='home'),
    url(r'^about/$', TemplateView.as_view(template_name='pages/about.html'), name='about'),
    # Your stuff: custom urls go here
]

urlpatterns += [

    url(r'^(?P<filename>(robots.txt)|(humans.txt))$',
        base_views.root_txt_files, name='root-txt-files'),

    # Rest API
    url(r'^api/', include(routers.router.urls)),

    # Django Admin
    url(r'^{}/'.format(settings.DJANGO_ADMIN_URL), admin.site.urls),

] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)

if settings.API_DEBUG:
    urlpatterns += [
        # Browsable API
        url('^schema/$', api_schemas.schema_view, name='schema'),
        url(r'^api-playground/$', api_schemas.swagger_schema_view, name='api-playground'),
        url(r'^api/auth-n/', include('rest_framework.urls', namespace='rest_framework')),
    ]

if settings.DEBUG:
    from django.views import defaults as dj_default_views
    from django.urls import get_callable

    urlpatterns += [
        url(r'^400/$', dj_default_views.bad_request, kwargs={'exception': Exception('Bad Request!')}),
        url(r'^403/$', dj_default_views.permission_denied, kwargs={'exception': Exception('Permission Denied!')}),
        url(r'^403_csrf/$', get_callable(settings.CSRF_FAILURE_VIEW)),
        url(r'^404/$', dj_default_views.page_not_found, kwargs={'exception': Exception('Not Found!')}),
        url(r'^500/$', handler500),
    ]

    # Django Debug Toolbar
    try:
        import debug_toolbar
        urlpatterns += [url(r'^__debug__/', include(debug_toolbar.urls)), ]
    except ImportError:
        pass


{{cookiecutter.github_repository}}/settings/common.py
"""Django settings for {{cookiecutter.project_name}} project.

see: https://docs.djangoproject.com/en/dev/ref/settings/
"""

# Third Party Stuff
{%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
import os
{%- endif %}
import environ
from django.utils.translation import ugettext_lazy as _


ROOT_DIR = environ.Path(__file__) - 2  # (/a/b/myfile.py - 2 = /a/)
APPS_DIR = ROOT_DIR.path('{{ cookiecutter.main_module }}')

env = environ.Env()

# INSTALLED APPS
# ==========================================================================
# List of strings representing installed apps.
# See: https://docs.djangoproject.com/en/dev/ref/settings/#installed-apps
INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django_sites',  # http://niwinz.github.io/django-sites/latest/
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.admin',
    # 'django.contrib.humanize',  # Useful template tags

    '{{ cookiecutter.main_module }}.base',
    '{{ cookiecutter.main_module }}.users',

    'rest_framework',  # http://www.django-rest-framework.org/
    'rest_framework_swagger',
    'versatileimagefield',  # https://github.com/WGBH/django-versatileimagefield/
{%- if cookiecutter.add_django_cors_headers.lower() == 'y' %}
    'corsheaders',   # https://github.com/ottoyiu/django-cors-headers/
{%- endif %}
{%- if cookiecutter.add_sass_with_django_compressor.lower() == 'y' %}

    'compressor',
{%- endif %}
{%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
    'raven.contrib.django.raven_compat',
{%- endif %}
    'mail_templated',  # https://github.com/artemrizhov/django-mail-templated
)

# INSTALLED APPS CONFIGURATION
# ==========================================================================

# django.contrib.auth
# ------------------------------------------------------------------------------
AUTH_USER_MODEL = 'users.User'
AUTHENTICATION_BACKENDS = ('django.contrib.auth.backends.ModelBackend',)

PASSWORD_HASHERS = [
    'django.contrib.auth.hashers.Argon2PasswordHasher',
    'django.contrib.auth.hashers.PBKDF2PasswordHasher',
    'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
    'django.contrib.auth.hashers.BCryptSHA256PasswordHasher',
    'django.contrib.auth.hashers.BCryptPasswordHasher',
]

AUTH_PASSWORD_VALIDATORS = [
    {'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator', },
    {'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator', 'OPTIONS': {'min_length': 6, }},
    {'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator', },
    {'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator', },
]

# For Exposing browsable api urls. By default urls won't be exposed.
API_DEBUG = env.bool('API_DEBUG', default=False)

# rest_framework
# ------------------------------------------------------------------------------
REST_FRAMEWORK = {
    'DEFAULT_PAGINATION_CLASS': '{{ cookiecutter.main_module }}.base.api.pagination.PageNumberPagination',
    'PAGE_SIZE': 30,

    # Default renderer classes for Rest framework
    'DEFAULT_RENDERER_CLASSES': [
        'rest_framework.renderers.JSONRenderer',
        'rest_framework.renderers.BrowsableAPIRenderer',
    ],

    # 'Accept' header based versioning
    # http://www.django-rest-framework.org/api-guide/versioning/
    'DEFAULT_VERSIONING_CLASS': 'rest_framework.versioning.AcceptHeaderVersioning',
    'DEFAULT_VERSION': '1.0',
    'ALLOWED_VERSIONS': ['1.0', ],
    'VERSION_PARAMETER': 'version',

    # Use Django's standard `django.contrib.auth` permissions,
    # or allow read-only access for unauthenticated users.
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.IsAuthenticated',
    ],
    'DEFAULT_THROTTLE_CLASSES': (
        'rest_framework.throttling.AnonRateThrottle',
    ),
    'DEFAULT_THROTTLE_RATES': {
        'anon': '10000/day',
    },
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.BasicAuthentication',

        # Primary api authentication
        '{{ cookiecutter.main_module }}.users.auth.backends.UserTokenAuthentication',

        # Mainly used for api debug.
        'rest_framework.authentication.SessionAuthentication',
    ),
    'EXCEPTION_HANDLER': '{{ cookiecutter.main_module }}.base.exceptions.exception_handler',
}

# https://django-rest-swagger.readthedocs.io/en/latest/settings/
SWAGGER_SETTINGS = {
    'LOGIN_URL': 'rest_framework:login',
    'LOGOUT_URL': 'rest_framework:logout',
    'SECURITY_DEFINITIONS': {
        # For BasicAuthentication
        'basic': {
            'type': 'basic'
        },
        # For UserTokenAuthentication
        "api_key": {
            "type": "apiKey",
            "name": "Authorization",
            "in": "header"
        },
    },
}

# DJANGO_SITES
# ------------------------------------------------------------------------------
# see: http://django-sites.readthedocs.org
SITE_SCHEME = env("SITE_SCHEME", default='http')
SITE_DOMAIN = env("SITE_DOMAIN", default='localhost:8000')
SITE_NAME = env("SITE_NAME", default='{{ cookiecutter.project_name }}')

# This is used in-case of the frontend is deployed at a different url than this django app.
FRONTEND_SITE_SCHEME = env('FRONTEND_SITE_SCHEME', default='https')
FRONTEND_SITE_DOMAIN = env('FRONTEND_SITE_DOMAIN', default='example.com')
FRONTEND_SITE_NAME = env('FRONTEND_SITE_NAME', default='{{ cookiecutter.project_name }}')

SITES = {
    'current': {
        'domain': SITE_DOMAIN,
        'scheme': SITE_SCHEME,
        'name': SITE_NAME
    },
    'frontend': {
        'domain': FRONTEND_SITE_DOMAIN,
        'scheme': FRONTEND_SITE_SCHEME,
        'name': FRONTEND_SITE_NAME
    },
}
SITE_ID = 'current'

# see user.services.send_password_reset
# password-confirm path should have placeholder for token
FRONTEND_URLS = {
    'home': '/',
    'password-confirm': '/reset-password/{token}/',
}

# MIDDLEWARE CONFIGURATION
# ------------------------------------------------------------------------------
# List of middleware classes to use.  Order is important; in the request phase,
# this middleware classes will be applied in the order given, and in the
# response phase the middleware will be applied in reverse order.
MIDDLEWARE = [
{%- if cookiecutter.add_django_cors_headers.lower() == 'y' %}
    'corsheaders.middleware.CorsMiddleware',
{%- endif %}
    'log_request_id.middleware.RequestIDMiddleware',  # For generating/adding Request id for all the logs
    'django.middleware.security.SecurityMiddleware',
{%- if cookiecutter.enable_whitenoise.lower() == 'y' %}
    'whitenoise.middleware.WhiteNoiseMiddleware',
{%- endif %}
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

# DJANGO CORE
# ------------------------------------------------------------------------------

# See: https://docs.djangoproject.com/en/dev/ref/settings/#debug
# Defaults to false, which is safe, enable them only in development.
DEBUG = env.bool('DJANGO_DEBUG', False)

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = '{{ cookiecutter.timezone }}'

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

# Languages we provide translations for
LANGUAGES = (
    ('en', _('English')),
)

if USE_TZ:
    # Add timezone information to datetime displayed.
    # https://mounirmesselmeni.github.io/2014/11/06/date-format-in-django-admin/
    from django.conf.locale.en import formats as en_formats
    en_formats.DATETIME_FORMAT = 'N j, Y, P (e)'

# A tuple of directories where Django looks for translation files.
LOCALE_PATHS = (
    str(APPS_DIR.path('locale')),
)

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# The list of directories to search for fixtures
# See: https://docs.djangoproject.com/en/dev/ref/settings/#std:setting-FIXTURE_DIRS
FIXTURE_DIRS = (
    str(APPS_DIR.path('fixtures')),
)

# The Python dotted path to the WSGI application that Django's internal servers
# (runserver, runfcgi) will use. If `None`, the return value of
# 'django.core.wsgi.get_wsgi_application' is used, thus preserving the same
# behavior as previous versions of Django. Otherwise this should point to an
# actual WSGI application object.
# See: https://docs.djangoproject.com/en/dev/ref/settings/#wsgi-application
WSGI_APPLICATION = 'wsgi.application'

# URL CONFIGURATION
# ------------------------------------------------------------------------------
ROOT_URLCONF = '{{ cookiecutter.main_module }}.urls'


# Use this to change base url path django admin
DJANGO_ADMIN_URL = env.str('DJANGO_ADMIN_URL', default='admin')

# EMAIL CONFIGURATION
# ------------------------------------------------------------------------------
EMAIL_BACKEND = env('DJANGO_EMAIL_BACKEND',
                    default='django.core.mail.backends.smtp.EmailBackend')

# DATABASE CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#databases
DATABASES = {
    'default': env.db('DATABASE_URL', default='postgres://postgres:postgres@localhost:5432/{{ cookiecutter.main_module }}'),
}
DATABASES['default']['ATOMIC_REQUESTS'] = True
DATABASES['default']['CONN_MAX_AGE'] = 10
{% if cookiecutter.postgis.lower() == 'y' %}DATABASES['default']['ENGINE'] = 'django.contrib.gis.db.backends.postgis'{% endif %}

# TEMPLATE CONFIGURATION
# -----------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#templates
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            str(APPS_DIR.path('templates')),
        ],
        'OPTIONS': {
            'debug': DEBUG,
            'loaders': [
                'django.template.loaders.filesystem.Loader',
                'django.template.loaders.app_directories.Loader',
            ],
            # See: https://docs.djangoproject.com/en/dev/ref/settings/#template-context-processors
            'context_processors': [
                '{{cookiecutter.main_module}}.base.context_processors.site_settings',
                'django.contrib.auth.context_processors.auth',
                'django.template.context_processors.debug',
                'django.template.context_processors.i18n',
                'django.template.context_processors.media',
                'django.template.context_processors.static',
                'django.template.context_processors.tz',
                'django.contrib.messages.context_processors.messages',
                'django.template.context_processors.request',
                # Your stuff: custom template context processors go here
            ],
        },
    },
]

CSRF_FAILURE_VIEW = '{{ cookiecutter.main_module }}.base.views.csrf_failure'

# STATIC FILE CONFIGURATION
# -----------------------------------------------------------------------------
# Absolute path to the directory static files should be collected to.
# Example: '/var/www/example.com/static/'
# See: https://docs.djangoproject.com/en/dev/ref/settings/#static-root
STATIC_ROOT = str(ROOT_DIR.path('.staticfiles'))

# URL that handles the static files served from STATIC_ROOT.
# Example: 'http://example.com/static/', 'http://static.example.com/'
# See: https://docs.djangoproject.com/en/dev/ref/settings/#static-url
STATIC_URL = '/static/'

# A list of locations of additional static files
STATICFILES_DIRS = (
    str(APPS_DIR.path('static')),
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
{%- if cookiecutter.add_sass_with_django_compressor.lower() == 'y' %}
    'compressor.finders.CompressorFinder',
{%- endif %}
)
{%- if cookiecutter.add_sass_with_django_compressor.lower() == 'y' %}

# Django Compressor Configuration
COMPRESS_CSS_FILTERS = [
    'django_compressor_autoprefixer.AutoprefixerFilter',
    'compressor.filters.cssmin.CSSMinFilter',
]

COMPRESS_PRECOMPILERS = (
    ('text/x-scss', 'django_libsass.SassCompiler'),
)

COMPRESS_ENABLED = True
COMPRESS_OFFLINE = env.bool('COMPRESS_OFFLINE', default=False)
{%- endif %}

# MEDIA CONFIGURATION
# ------------------------------------------------------------------------------

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: '/var/www/example.com/media/'
# See: https://docs.djangoproject.com/en/dev/ref/settings/#media-root
MEDIA_ROOT = str(ROOT_DIR.path('.media'))

# URL that handles the media served from MEDIA_ROOT.
# Examples: 'http://example.com/media/', 'http://media.example.com/'
# See: https://docs.djangoproject.com/en/dev/ref/settings/#media-url
MEDIA_URL = env("MEDIA_URL",
                default="{}://{}/media/".format(SITE_SCHEME, SITE_DOMAIN))

#  SECURITY
# -----------------------------------------------------------------------------
CSRF_COOKIE_HTTPONLY = False  # Allow javascripts to read CSRF token from cookies
SESSION_COOKIE_HTTPONLY = True  # Do not allow Session cookies to be read by javascript

SECURE_CONTENT_TYPE_NOSNIFF = True
SECURE_BROWSER_XSS_FILTER = True
X_FRAME_OPTIONS = 'DENY'

# django-log-request-id - Sending request id in response
REQUEST_ID_RESPONSE_HEADER = 'REQUEST_ID'

{%- if cookiecutter.add_celery.lower() == 'y' %}

# DJANGO CELERY CONFIGURATION
# -----------------------------------------------------------------------------
# see: http://celery.readthedocs.org/en/latest/userguide/tasks.html#task-states
CELERY_BROKER_URL = env('REDIS_URL', default="redis://localhost:6379/0")
CELERY_ACCEPT_CONTENT = ['json']
CELERY_TASK_SERIALIZER = 'json'
CELERY_RESULT_SERIALIZER = 'json'
CELERY_TIMEZONE = env('CELERY_TIMEZONE', default=TIME_ZONE)  # Use django's timezone by default
{%- endif %}

# EMAIL
# ------------------------------------------------------------------------------
DEFAULT_FROM_EMAIL = env('DEFAULT_FROM_EMAIL',
                         default='{{ cookiecutter.default_from_email }}')
EMAIL_SUBJECT_PREFIX = env('EMAIL_SUBJECT_PREFIX', default='[{{cookiecutter.project_name}}] ')
EMAIL_USE_TLS = env.bool('EMAIL_USE_TLS', default=True)
SERVER_EMAIL = env('SERVER_EMAIL', default=DEFAULT_FROM_EMAIL)

# LOGGING CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#logging
# Default logging for Django. This sends an email to the site admins on every
# HTTP 500 error. Depending on DEBUG, all other log records are either sent to
# the console (DEBUG=True) or discarded by mean of the NullHandler (DEBUG=False).
# See http://docs.djangoproject.com/en/dev/topics/logging

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        },
        'request_id': {
            '()': 'log_request_id.filters.RequestIDFilter'
        }
    },
    'formatters': {
        'complete': {
            # NOTE: make sure to include 'request_id' in filters when using this
            # formatter in any handlers.
            'format': '%(asctime)s:[%(levelname)s]:logger=%(name)s:request_id=%(request_id)s message="%(message)s"'
        },
        'simple': {
            'format': '%(levelname)s:%(asctime)s: %(message)s'
        },
        'null': {
            'format': '%(message)s',
        },
    },
    'handlers': {
        'null': {
            'level': 'DEBUG',
            'class': 'logging.NullHandler',
        },
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'complete',
            'filters': ['request_id'],
        },
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        },
        {%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
        'sentry': {
            'level': 'ERROR',
            'class': 'raven.contrib.django.raven_compat.handlers.SentryHandler',
            'formatter': 'complete',
            'filters': ['request_id'],
        },
        {%- endif %}
    },
    'loggers': {
        'django': {
            'handlers': ['null'],
            'propagate': False,
            'level': 'INFO',
        },
        'django.request': {
            'handlers': ['mail_admins', 'console'],
            'level': 'ERROR',
            'propagate': False,
        },
        'django.server': {
            'handlers': ['console'],
            'level': 'INFO',
            'propagate': False,
        },
        '{{cookiecutter.main_module}}': {
            'handlers': ['console'],
            'level': 'INFO',
            'propagate': False,
        },
        # Catch All Logger -- Captures any other logging
        '': {
            'handlers': [
                'console',
                {%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
                'sentry'
                {%- endif %}
            ],
            'level': 'ERROR',
            'propagate': True,
        },
    }
}


def get_release():
    import {{cookiecutter.main_module}}
    {%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
    import raven
    from raven import exceptions as raven_exceptions
    {%- endif %}
    release = {{cookiecutter.main_module}}.__version__
    {%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
    try:
        git_hash = raven.fetch_git_sha(os.path.dirname(os.pardir))[:7]
        release = '{}-{}'.format(release, git_hash)
    except raven_exceptions.InvalidGitRepository:
        pass
    {%- endif %}
    return release


RELEASE_VERSION = get_release()


{%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
RAVEN_CONFIG = {
    'dsn': env('SENTRY_DSN', default=''),
    'environment': env('SENTRY_ENVIRONMENT', default='production'),
    'release': RELEASE_VERSION,
}
{%- endif %}

SITE_INFO = {
    'RELEASE_VERSION': RELEASE_VERSION,
{%- if cookiecutter.use_sentry_for_error_reporting == 'y' %}
    'IS_RAVEN_INSTALLED': RAVEN_CONFIG['dsn'] is not ''
{%- endif %}
}
